File: Top_Interview_Questions/Hard/23. Merge k Sorted Lists/solution5.cpp
#include <vector>
using namespace std;

/**
 * 參考解答: 用 divide and conquer 來做，每次都切一半，直到剩下兩個 list 的時候
 * 就 merge，最後全部合併起來就是答案
 * 時間: O(mlogm), m: n1+n2+...+nk, 空間: O(k)
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr)
    {
    }
    ListNode(int x) : val(x), next(nullptr)
    {
    }
    ListNode(int x, ListNode *next) : val(x), next(next)
    {
    }
};

class Solution
{
  public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.size() == 0)
        {
            return nullptr;
        }
        return merge(lists, 0, lists.size() - 1);
    }

    ListNode *merge(vector<ListNode *> &lists, int left, int right)
    {
        if (left == right)
        {
            return lists[left];
        }
        int mid = (left + right) / 2;
        ListNode *list1 = merge(lists, left, mid);
        ListNode *list2 = merge(lists, mid + 1, right);
        return mergeTwoLists(list1, list2);
    }

    ListNode *mergeTwoLists(ListNode *left, ListNode *right)
    {
        ListNode *dummy = new ListNode(), *result = dummy;
        while (left != nullptr && right != nullptr)
        {
            if (left->val < right->val)
            {
                result->next = left;
                left = left->next;
            }
            else
            {
                result->next = right;
                right = right->next;
            }
            result = result->next;
        }
        result->next = left != nullptr ? left : right;
        return dummy->next;
    }
};

File: Top_Interview_Questions/Hard/23. Merge k Sorted Lists/solution5.py
from typing import List, Optional


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None
        return self.merge(lists, 0, len(lists) - 1)

    def merge(self, lists: List[Optional[ListNode]], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return lists[left]
        mid = int((left + right) / 2)
        list1 = self.merge(lists, left, mid)
        list2 = self.merge(lists, mid + 1, right)
        return self.merge_two_lists(list1, list2)

    def merge_two_lists(self, left: Optional[ListNode], right: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        result = dummy
        while left is not None and right is not None:
            if left.val < right.val:
                result.next = left
                left = left.next
            else:
                result.next = right
                right = right.next
            result = result.next
        result.next = left if left is not None else right
        return dummy.next
